File: backends/rails/app/services/forms_service.rb
module FormsService
  extend self

  def current_forms
    prev_published = Form.where(published: true, current: false).pluck(:number)
    select_str = prev_published.any? ? "number IN (#{prev_published.join(',')})" : "FALSE"
    Form
      .joins("LEFT OUTER JOIN form_responses ON forms.id = form_responses.form_id")
      .includes(:user, :fields)
      .select("forms.*, #{select_str} as prev_published, count(form_responses.*) as num_responses")
      .where(current: true)
      .group("forms.id")
  end

  def create(attrs)
    attrs[:number] ||= Form.max_number + 1
    attrs[:version] = Form.max_version(attrs[:number]) + 1
    attrs[:fields] = build_fields(attrs[:fields])
    Form.where(number: attrs[:number]).update_all(current: false)
    Form.create!(filter_create_form_attrs(attrs).merge(current: true))
  end

  def versions(number)
    forms = Form
      .joins("LEFT OUTER JOIN form_responses ON forms.id = form_responses.form_id")
      .includes(:user, :fields)
      .select("forms.*, count(form_responses.*) as num_responses")
      .where(number: number)
      .group("forms.id")
    raise ActiveRecord::RecordNotFound if forms.length == 0
    forms
  end

  def version(number, version)
    Form.includes(:user, :fields).find_by!(number: number, version: version)
  end

  def update(number, version, attrs)
    form = Form.find_by!(number: number, version: version)
    raise Form::Locked if form.locked
    form.fields.destroy_all
    attrs[:fields] = build_fields(attrs[:fields])
    form.update!(filter_update_form_attrs(attrs))
    form
  end

  def publish(number, version)
    Form.where(number: number).update_all(published: false)
    form = Form.find_by!(number: number, version: version)
    form.update!(slug: make_unique_slug(form), published: true, locked: true)
    form
  end

  def unpublish(number, version)
    form = Form.find_by!(number: number, version: version)
    form.update!(published: false, slug: '')
    form
  end

  def form_with_responses(number, version)
    Form
      .includes(:user, :fields, responses: [:user, :field_responses])
      .find_by!(number: number, version: version)
  end

  def published_form(project, slug)
    Form
      .includes(:user, :fields)
      .find_by!(published: true, project: project, slug: slug)
  end

  def published_forms(project)
    Form.includes(:user, :fields).where(published: true, project: project)
  end

  def build_response(form, details_attrs, user = nil)
    form_response = form.responses.build(user: user)
    non_info_fields(form.fields).each do |field|
      attrs = {form_field: field, details: details_attrs.fetch(field.id.to_s, {})}
      form_response.field_responses.build(attrs)
    end
    form_response
  end

  def create_response(form, details_attrs, user = nil)
    form_response = build_response(form, details_attrs, user)
    form_response.save!
    form_response
  end

  private

  def filter_create_form_attrs(attrs)
    allowed_attrs = [:number, :version, :name, :description, :project, :user, :fields]
    attrs.select { |k, v| allowed_attrs.include?(k.to_sym) }
  end

  def filter_update_form_attrs(attrs)
    allowed_attrs = [:name, :description, :project, :user, :fields]
    attrs.select { |k, v| allowed_attrs.include?(k.to_sym) }
  end

  def build_fields(fields)
    (fields || []).map do |field_attrs|
      FormField.new(filter_field_attrs(field_attrs))
    end
  end

  def filter_field_attrs(attrs)
    allowed_attrs = [:kind, :details]
    attrs.select { |k, v| allowed_attrs.include?(k.to_sym) }
  end

  def non_info_fields(fields)
    fields.reject { |field| field.kind == 'info' }
  end

  def make_unique_slug(form)
    slug = form.name.parameterize
    used_slugs = Form.where("published = TRUE AND slug LIKE ?", "#{slug}%").pluck(:slug)
    if used_slugs.empty? or !used_slugs.include?(slug)
      slug
    else
      nums = used_slugs.map { |s| s.match(/^#{slug}-(\d+)$/) { |m| m[1].to_i } }.compact
      slug_num = (nums.max || 0) + 1
      "#{slug}-#{slug_num}"
    end
  end

end

File: backends/django/django_site/apps/forms_api/services/forms_service.py
import re

from django.core.exceptions import ValidationError
from django.db.models import Count
from django.utils.text import slugify

from django_site.apps.forms_api.models import (
    Form, FormField, FormResponse, FormFieldResponse, FormLockedError
)


__all__ = ['current_forms', 'create']

ALLOWED_UPDATE_ATTRS = ['name', 'description', 'project', 'user', 'fields']
ALLOWED_CREATE_ATTRS = ['number', 'version'] + ALLOWED_UPDATE_ATTRS


def _filter_data(data, allowed):
    return {key: data[key] for key in allowed if key in data}


def _validate_and_save_form_data(form_data, form=None):
    fields = form_data.pop('fields', [])
    if form is None:
        form = Form(**form_data)
    else:
        for key, value in form_data.items():
            setattr(form, key, value)
    form.full_clean()
    form.save()
    return _validate_and_save_fields(form, fields)


def _validate_and_save_fields(form, fields):
    for field_data in fields:
        field_data['form'] = form
        field = FormField(**field_data)
        field.full_clean()
        field.save()
    return form


def _make_unique_slug(form):
    slug = slugify(form.name)
    used_slugs = (Form.objects
                  .filter(published=True, slug__startswith=slug)
                  .values_list('slug', flat=True))
    if slug in used_slugs:
        slug_matches = [re.match('%s-(\d+)$' % slug, s) for s in used_slugs]
        slug_nums = [int(m.group(1)) for m in slug_matches if m]
        slug_num = (max(slug_nums) if slug_nums else 0) + 1
        slug = '%s-%d' % (slug, slug_num)
    return slug


def current_forms():
    prev_pub = Form.objects.filter(published=True, current=False)
    numbers = [str(n) for n in prev_pub.values_list('number', flat=True)]
    prev_pub_sql = ("number IN (%s)" % ','.join(numbers) if len(numbers) > 0
                    else "FALSE")
    return (Form.objects.filter(current=True)
            .annotate(num_responses=Count('formresponse'))
            .extra(select={'prev_published': prev_pub_sql}))


def create(**form_data):
    form_data = _filter_data(form_data, ALLOWED_CREATE_ATTRS)
    if 'number' in form_data and form_data['number'] > 0:
        next_version = Form.objects.max_version(form_data['number']) + 1
        form_data['version'] = next_version
    else:
        form_data['number'] = Form.objects.max_number() + 1
        form_data['version'] = 1
    form_data['current'] = True
    Form.objects.filter(number=form_data['number']).update(current=False)
    return _validate_and_save_form_data(form_data)


def update(number, version, **form_data):
    form_data = _filter_data(form_data, ALLOWED_UPDATE_ATTRS)
    form = Form.objects.get(number=number, version=version)
    if form.locked:
        raise FormLockedError
    FormField.objects.filter(form=form).delete()
    return _validate_and_save_form_data(form_data, form)


def versions(number):
    forms = (Form.objects
             .select_related('user')
             .prefetch_related('formfield_set')
             .filter(number=number)
             .annotate(num_responses=Count('formresponse')))
    if not forms:
        raise Form.DoesNotExist
    return forms


def version(number, version):
    return (Form.objects
            .select_related('user')
            .prefetch_related('formfield_set')
            .get(number=number, version=version))


def publish(number, version):
    Form.objects.filter(number=number).update(published=False)
    form = Form.objects.get(number=number, version=version)
    form.published = True
    form.locked = True
    form.slug = _make_unique_slug(form)
    form.save()
    return form


def unpublish(number, version):
    form = Form.objects.get(number=number, version=version)
    form.published = False
    form.save()
    return form


def form_with_responses(number, version):
    return (Form.objects
            .prefetch_related('formfield_set',
                              'formresponse_set__formfieldresponse_set')
            .get(number=number, version=version))


def published_form(project, slug):
    return (Form.objects
            .select_related('user')
            .prefetch_related('formfield_set')
            .get(project=project, slug=slug, published=True))


def published_forms(project):
    return (Form.objects
            .select_related('user')
            .prefetch_related('formfield_set')
            .filter(project=project, published=True))


def create_response(form, field_response_data, user=None):
    fields = filter(lambda f: f.kind != 'info', form.formfield_set.all())
    field_responses = _build_field_responses(fields, field_response_data)
    _validate_field_responses(field_responses)
    form_response = FormResponse.objects.create(form=form, user=user)
    for field_response in field_responses:
        field_response.form_response = form_response
        field_response.save()
    return form_response


def _build_field_responses(fields, field_response_data):
    field_responses = []
    for field in fields:
        details = field_response_data.get(str(field.id), {})
        field_response = FormFieldResponse(form_field=field, details=details)
        field_responses.append(field_response)
    return field_responses


def _validate_field_responses(field_responses):
    errors = {}
    for field_response in field_responses:
        try:
            field_response.clean()
        except ValidationError as e:
            errors[str(field_response.form_field.id)] = e
    if errors:
        raise ValidationError(errors)
